From 4510ecad111c87fc822282e4cb1eb6ac6274dd4e Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:58 -0400
Subject: [virt] kvm/nvmx: Implement support for EFER saving on VM-exit

Message-id: <1406040016-3289-84-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86163
O-Subject: [RHEL7 PATCH v2 083/221] KVM: nVMX: Implement support for EFER saving on VM-exit
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

Implement and advertise VM_EXIT_SAVE_IA32_EFER. L0 traps EFER writes
unconditionally, so we always find the current L2 value in the
architectural state.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Reviewed-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 10ba54a58949e222dcb9cd9fb64cd3e7f53fc6b1)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 862a465..2b4602d 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -2206,7 +2206,7 @@ static __init void nested_vmx_setup_ctls_msrs(void)
 #endif
   VM_EXIT_LOAD_IA32_PAT | VM_EXIT_SAVE_IA32_PAT;
  nested_vmx_exit_ctls_high |= (VM_EXIT_ALWAYSON_WITHOUT_TRUE_MSR |
-          VM_EXIT_LOAD_IA32_EFER);
+  VM_EXIT_LOAD_IA32_EFER | VM_EXIT_SAVE_IA32_EFER);
 
  /* entry controls */
  rdmsr(MSR_IA32_VMX_ENTRY_CTLS,
@@ -8131,6 +8131,8 @@ static void prepare_vmcs12(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12)
  vmcs12->guest_ia32_debugctl = vmcs_read64(GUEST_IA32_DEBUGCTL);
  if (vmcs12->vm_exit_controls & VM_EXIT_SAVE_IA32_PAT)
   vmcs12->guest_ia32_pat = vmcs_read64(GUEST_IA32_PAT);
+ if (vmcs12->vm_exit_controls & VM_EXIT_SAVE_IA32_EFER)
+  vmcs12->guest_ia32_efer = vcpu->arch.efer;
  vmcs12->guest_sysenter_cs = vmcs_read32(GUEST_SYSENTER_CS);
  vmcs12->guest_sysenter_esp = vmcs_readl(GUEST_SYSENTER_ESP);
  vmcs12->guest_sysenter_eip = vmcs_readl(GUEST_SYSENTER_EIP);
-- 
1.7.1